### PR TITLE
CORE-2003: Migrate composite UI components (Tooltip, Toast, Tabs, etc.)

### PR DESCRIPTION
**Pull request by OpenStaxClaude**
## Summary

Migrated composite UI components from styled-components to plain CSS with CSS variables as part of the migration away from styled-components.

### Components migrated:
- ✅ Tooltip (including TooltipGroup and CustomTooltip)
- ✅ Toast
- ✅ ToastContainer (including BodyPortalToastContainer)
- ✅ Tabs (including button-bar variant)
- ✅ ButtonBar
- ✅ DropdownMenu (already migrated, included for completeness)

### Migration approach:
- Converted styled-components to className-based styling
- Created separate CSS files for each component
- Bound theme values to CSS variables for consistency
- Used classnames library for conditional styling
- Maintained all existing functionality, variants, and props

### CSS Variables used:
Each component now uses CSS custom properties that reference the palette:
- Tooltip: `--tooltip-bg`, `--tooltip-color`, `--tooltip-border-color`
- Toast: `--toast-success-*`, `--toast-neutral-*`, `--toast-failure-*`
- ToastContainer: `--toast-container-z-index`
- Tabs: `--tabs-border-color`, `--tabs-active-border-color`, `--tabs-button-*`
- ButtonBar: `--button-bar-border-color`, `--button-bar-selected-bg`, `--button-bar-hover-bg`

### Testing
All components have existing Ladle stories that demonstrate:
- Different size variants (small, medium, large)
- Different visual variants (button-bar, success/neutral/failure)
- Interactive states (hover, selected)
- Different placements (for Tooltip)

### Related
- Jira: [CORE-2003](https://openstax.atlassian.net/browse/CORE-2003)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[CORE-2003]: https://openstax.atlassian.net/browse/CORE-2003?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's unresolved comment

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's unresolved comment

**Comment by OpenStaxClaude** [comment]
Addressed the code review comment about `ButtonBarProps` breaking API change.

### Changes made:
- Extended `ButtonBarProps` with `React.HTMLAttributes<HTMLDivElement>` to allow passing standard div attributes (className, style, id, etc.)
- Properly destructured and merged `className` prop with component-generated classes using classnames
- Properly destructured and merged custom `style` prop with CSS variable definitions

This maintains backward compatibility while allowing callers to customize the component with standard HTML attributes as expected.

Commit: 05706448

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's latest comments.

**Comment by OpenStaxClaude** [comment]
Addressed all 3 Copilot code review comments from the latest review.

## 1. Fixed Tabs style prop merge order (src/components/Tabs.tsx)

**Issue:** Consumer-provided CSS variables were being overridden by component defaults, making it impossible to customize `--tabs-*` values.

**Fix:** Changed style merge order from `{...style, ...additionalStyle}` to `{...additionalStyle, ...style}`. Now consumer-provided styles take precedence.

## 2. Added style prop support to BodyPortal (src/components/BodyPortal.tsx & ToastContainer.tsx)

**Issue:** `BodyPortalToastContainer` couldn't set CSS variables because `BodyPortal` didn't apply the `style` prop to the portal element.

**Fix:**
- Added `style?: React.CSSProperties` to `BodyPortalProps`
- Applied style to portal element using `Object.assign(element.style, style)` in useLayoutEffect
- Added cleanup to remove styles on unmount
- Updated `BodyPortalToastContainer` to pass `style` prop with z-index CSS variable

This fixes the z-index theme binding for toast containers.

## 3. Removed unused icon prop from Tooltip (src/components/Tooltip.tsx)

**Issue:** The `icon` prop was destructured in `Tooltip` but never used (only used in `TooltipGroup`).

**Fix:** Removed `icon` from the destructuring in `Tooltip` component to eliminate the unused variable.

All changes maintain backward compatibility while fixing the identified issues.

Commit: 67056440

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's latest comments

**Comment by OpenStaxClaude** [comment]
Addressed all 3 Copilot code review comments from the latest review.

## 1. Fixed BodyPortal style cleanup (src/components/BodyPortal.tsx:95-99)

**Issue:** Cleanup was using `removeProperty(key)` with camelCase property names (e.g., `backgroundColor`), but `removeProperty` expects kebab-case (e.g., `background-color`). This caused non-CSS-variable styles to leak on unmount.

**Fix:** Added conversion from camelCase to kebab-case before calling `removeProperty`:
```typescript
const kebabKey = key.replace(/([A-Z])/g, '-$1').toLowerCase();
element.style.removeProperty(kebabKey);
```

## 2. Memoized style object in BodyPortalToastContainer (src/components/ToastContainer.tsx:50-56)

**Issue:** Fresh style object created on each render caused `BodyPortal`'s `useLayoutEffect` to trigger unnecessarily, potentially tearing down and re-inserting the portal element when toast list changes.

**Fix:** Wrapped style object in `React.useMemo` with empty dependency array:
```typescript
const style = React.useMemo(() => ({
  '--toast-container-z-index': zIndex.toasts,
} as React.CSSProperties), []);
```

## 3. Prevented invalid props from reaching AriaTooltip (src/components/Tooltip.tsx:8-16)

**Issue:** `icon` and `ariaLabel` props are trigger-only concerns (used by `TooltipGroup`), but were being spread to `AriaTooltip`, potentially causing invalid DOM attributes or unexpected react-aria behavior.

**Fix:** Explicitly destructured `icon` and `ariaLabel` to prevent them from being spread:
```typescript
export const Tooltip = ({children, placement, icon, ariaLabel, ...props}: ...) => {
  // icon and ariaLabel are now excluded from ...props
  return <AriaTooltip {...props} placement={placement} ... />
}
```

All fixes maintain backward compatibility while addressing the performance and correctness issues identified.

Commit: f9661689

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's latest comments

**Comment by OpenStaxClaude** [comment]
Addressed both code review comments about `BodyPortal`:

## 1. Fixed performance issue with style dependency (Comment 2)

**Issue:** Including `style` in the useLayoutEffect dependency array caused the portal element to be removed and re-inserted on every style change. This could lead to focus loss or UI flicker, especially problematic since callers often pass a new style object literal each render.

**Fix:** Split the single useLayoutEffect into two separate effects:
- **Portal creation effect**: Handles creating/destroying the portal element and setting non-style props (className, id, role, etc.). Does NOT depend on `style`.
- **Style update effect**: Only handles applying/removing styles. Depends on `style` but updates styles in-place without removing the portal element from the DOM.

This approach:
- Eliminates unnecessary DOM churn when only styles change
- Prevents focus loss and flicker
- Works correctly even when callers don't memoize the style prop (though memoization is still recommended for performance)

## 2. Added comprehensive test coverage (Comment 1)

Added 6 new test cases to `BodyPortal.spec.tsx`:
- ✅ Applies CSS variables from style prop
- ✅ Applies regular CSS properties from style prop  
- ✅ Updates styles when style prop changes **without removing portal element** (validates the performance fix)
- ✅ Removes styles on unmount
- ✅ Handles null and undefined style values gracefully

These tests ensure:
- CSS variables and regular styles are applied correctly
- The portal element stays in the DOM during style updates (performance)
- Styles are properly cleaned up on unmount (no leaks)
- Edge cases like null/undefined values are handled

Commit: d6c29829

**Comment by OpenStaxClaude** [comment]
@RoyEJohnson Rebased onto `main` and resolved the conflicts — the branch is now 0 behind / 7 ahead, so it fast-forwards cleanly. I left the Jira check alone as you said.

**Conflicts and how I resolved them**

- **`Tabs.tsx`** — took the migrated (CSS) version and carried #135's forced-colors selected-tab styles into `Tabs.css`. I also added them to `ButtonBar.css`: on `main` `ButtonBar` inherited that fix implicitly by interpolating `buttonBarItemCss` from `Tabs`, and once the CSS is split per component that link is gone, so it has to be declared in both files.
- **`BodyPortal.tsx` / `BodyPortal.spec.tsx`** — took `main`'s version wholesale. `main` landed the same CSS-variable support via `CSSPropertiesWithVariables` in `src/types`, which is the better-typed version of what this branch had.
- **`ButtonBar` / `Radio` / `Tabs` snapshots** — took `main`'s, then regenerated with `jest -u`. The only diffs are the expected `sc-*` → semantic-class swap plus the inline CSS-variable `style` attribute.

**Fallout from the rebase, plus the outstanding Copilot note**

- Switched the CSS-variable style objects to `CSSPropertiesWithVariables` instead of casting to `React.CSSProperties`. This wasn't cosmetic — `BodyPortal` requires that type on `main`, so `ToastContainer` didn't typecheck otherwise.
- Narrowed `Tooltip`'s props so `icon`/`ariaLabel` are only accepted by `TooltipGroup`, which is the component that actually owns the trigger button. That's the low-confidence Copilot comment from the 7/22 review; it was a real (if minor) point, so I took it.
- `CustomTooltip` now merges a caller-supplied `className`/`style` instead of clobbering them. styled-components used to do that merging for us and the first pass of the migration dropped it.

**One thing I'd like your call on**

`palette.darkerGreen` is `"4e7226"` — no leading `#`. On `main` that means the success toast title renders `color: 4e7226`, which is invalid CSS and silently ignored, so the title has never actually been green. The migration normalizes it to `#4e7226`, so success toast titles now render green for the first time. I kept the normalization local to `Toast.tsx` rather than touching the exported palette value, since `colors.palette` is part of the public surface. Happy to instead fix `palette.darkerGreen` properly, or preserve the old broken behavior — whichever you prefer.

**Verification:** `npm run lint`, `npx tsc --noEmit`, `npm run test` (39 suites / 231 tests), `npm run build`, and `npm run ladle-build` all pass locally, and `build` is green on CI.

File: src/components/Tabs.tsx [hunk 1]
```diff
@@ -1,140 +1,48 @@
...
+  };
+
   return (
-    <StyledTabs
-      variant={variant}
-      size={size}
-      className={className}
+    <RAC.Tabs
+      className={tabsClass}
+      style={{...additionalStyle, ...style}}
```
[the remark this answers is not in the document]
Comment (OpenStaxClaude) [reply src/components/Tabs.tsx]: Confirmed and fixed in 2bd2044.

The root cause is that `TabsProps` **intersected** `RAC.TabsProps` with static `className`/`style`/`children`, and an intersection collapses react-aria's `T | ((values) => T)` union down to `T` — `string & ((values) => string)` is just `string`. So render callbacks were rejected at compile time, and the `{...style}` spread turned a callback into `{}` at runtime.

The fix drops the narrowing entirely and resolves each callback against the render props before merging:

```tsx
className={(values) => classNames(
  variantClassName,
  typeof className === 'function' ? className(values) : className
)}
style={(values) => ({
  ...cssVariables,
  ...(typeof style === 'function' ? style(values) : style),
})}
```

Two things worth noting:

- `children` was narrowed the same way, so the children render-prop form was broken too. It now flows through untouched.
- I checked `useRenderProps` in react-aria-components — it already does `{...defaultStyle, ...computedStyle}` whether `style` is a callback or an object, so merging `defaultStyle` here would be redundant. We only merge the caller's.

Added two regression tests in `Tabs.spec.tsx`. Against the previous code the callback one fails to compile with exactly the union-collapse you described:

```
error TS2322: Type '({ orientation }) => string' is not assignable to type
'string | (string & ((values) => string)) | undefined'.
```

The existing snapshots are unchanged, since a callback returning a string produces the same `class` attribute as passing that string directly.

File: src/components/Tooltip.tsx [hunk 2]
```diff
@@ -1,116 +1,76 @@
...
-export const Tooltip = ({children, placement, icon, ...props}: React.PropsWithChildren<TooltipProps>) => 
-  <StyledTooltip {...props} placement={placement}>
-  <OverlayArrow>
-    <svg width={8} height={8} viewBox="0 0 8 8">
-      <path d="M0 0 L4 4 L8 0" stroke="#ccc" strokeWidth="1" />
-    </svg>
-  </OverlayArrow>
-  {children}
-</StyledTooltip>;
+export const Tooltip = ({children, placement, ...props}: React.PropsWithChildren<TooltipProps>) => {
```
[the remark this answers is not in the document]
Comment (OpenStaxClaude) [reply src/components/Tooltip.tsx]: Good catch — fixed in 2bd2044 by retaining both as compatibility wrappers.

I searched the `openstax` org before deciding between a wrapper and a documented removal, and it's not theoretical for one of them:

- **`StyledTrigger`** — used by `openstax/assignments`: `LicenseNotice.tsx:22` does `<UI.StyledTrigger>{licenseText}</UI.StyledTrigger>`. Removing it breaks that build.
- **`StyledTooltip`** — no consumers outside this repo. The other org hits are `StyledTooltipGroup` (a local `styled(UI.TooltipGroup)` in assignments) and an unrelated `styled(react-bootstrap.Tooltip)` in tutor-js.

Both are back as plain-CSS components that apply `.tooltip` / `.tooltip-trigger` and merge a caller `className`/`style`, and both carry `@deprecated` pointing at `Tooltip`/`TooltipGroup`. They're typed to accept only the string/object forms of `className`/`style`, which is what the styled-components versions actually supported — better to reject a render callback at compile time than drop it silently.

**This also surfaced a second, unreported regression.** `Tooltip` was setting `className="tooltip"` *after* `{...props}`, so a caller's `className` was discarded. styled-components used to merge it, and assignments depends on that: `Settings/styles.ts:13` is `styled(UI.TooltipGroup)`, whose generated class flows `TooltipGroup` → `Tooltip` → the tooltip element. On this branch those styles would have silently stopped applying. Both now merge, with a test that fails on the old behavior.

For the remaining removals — the `tabListBaseCss` / `tabBaseCss` / `buttonBarWrapperCss` / `buttonBarItemCss` fragments — the same search found no consumers outside this repo, so I've taken the documentation route rather than shimming them. They're written up in `CHANGELOG.md` under the existing `[Unreleased]` → BREAKING CHANGES section, matching the format the CORE-1999 Button migration established. Note the precedent here is mixed: #130 kept semantic names like `NavBarButton` as plain components, while #120 dropped `StyledLabel`/`StyledInput` outright — so I've kept what's used and documented what isn't.

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Address Copilot's latest comment

File: src/components/Tooltip.tsx [hunk 3]
```diff
@@ -1,95 +1,73 @@
...
+/**
+ * @deprecated The styles now live in the `.tooltip` class in Tooltip.css. Prefer `Tooltip`;
+ * this remains so consumers that composed the old styled-component keep working.
+ */
+export const StyledTooltip = ({className, style, ...props}: Omit<AriaTooltipProps, 'className' | 'style'> & ClassNameAndStyle) =>
+  <AriaTooltip
+    {...props}
+    className={classNames('tooltip', className)}
+    style={{...tooltipCssVariables, ...style}}
+  />;
```
[the remark this answers is not in the document]
Comment (OpenStaxClaude) [reply src/components/Tooltip.tsx]: Correct on both counts — fixed in dcde66a.

I verified the premise before changing anything: styled-components has forwarded refs since v4 and this repo is on 5.3.5, so `styled(AriaTooltip)` and `styled(Button)` did pass a ref through. Both underlying react-aria components accept one — `Tooltip` is `RefAttributes<HTMLDivElement>`, `Button` is `RefAttributes<HTMLButtonElement>` — so the capability was real and my replacements dropped it.

Both are now wrapped in `forwardRef`, following the pattern the other migrated react-aria wrappers use (`NavBarButton`, `NavBarPopover` from #130), including `displayName`:

```tsx
export const StyledTrigger = React.forwardRef<
  React.ElementRef<typeof Button>,
  Omit<AriaButtonProps, 'className' | 'style'> & ClassNameAndStyle
>(({className, style, ...props}, ref) => (
  <Button ref={ref} className={classNames('tooltip-trigger', className)} style={style} {...props} />
));
StyledTrigger.displayName = 'StyledTrigger';
```

I took the ref types from `React.ElementRef<typeof ...>` rather than hardcoding the element, matching `NavBarButton`, so they track react-aria if it changes.

Added a test per export asserting the ref resolves to the real DOM node. Against the previous code both fail to compile with exactly the error you predicted:

```
error TS2322: Property 'ref' does not exist on type
'IntrinsicAttributes & Omit<ButtonProps, "className" | "style"> & ClassNameAndStyle'.
```

**Comment by OpenStaxClaude** [comment]
@RoyEJohnson Copilot's `forwardRef` comment is addressed in dcde66a — details in the thread. Both checks are green on `dcde66a`, including Jira now that 70af7d9 fixed the secret names.

One unrelated thing I ran into and want to flag rather than quietly leave. While verifying I hit an intermittent failure in `ProfileMenu › matches snapshot with user initials`, which this PR doesn't touch. It reproduces reliably with a **cold jest cache running the full suite**, and never when `ProfileMenu` is run on its own — so it's an ordering/parallelism interaction. The diff is always react-aria focus state:

```
-       data-focus-visible="true"
-       data-focused="true"
```

I checked out unmodified `origin/main` in a throwaway worktree and it fails there too, 4/4 cold-cache runs (231 tests, i.e. without my additions). So it's pre-existing on `main`, not something this branch introduced. Warm-cache runs pass, which is why CI stays green.

I've left it alone since it's outside this PR's scope, but it's a real latent flake — happy to open a ticket if you want it chased.

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Open a separate bug-fix ticket for the testing issue.

**Comment by OpenStaxClaude** [comment]
@RoyEJohnson Filed as **[CORE-2715: Fix flaky ProfileMenu snapshot test in ui-components](https://openstax.atlassian.net/browse/CORE-2715)** — Bug, component REX, `Maintenance`, Low priority, in Backlog.

It carries the reproduction command, the exact failing diff, and the evidence it's pre-existing. I re-verified against current `main` (`5fdc635`) before filing: 3/3 cold-cache full-suite runs fail there, so it's independent of this branch. The acceptance criteria explicitly rule out "fixes" that just re-record the snapshot or skip the test, since either would hide whatever is non-deterministic. Triage hypotheses about react-aria's input-modality state are in a collapsed block, labelled as unconfirmed rather than presented as a diagnosis.

Two things to flag:

- The API rejected my `relates to` link to CORE-2003 with `401 No Link Issue Permission`, so I recorded the relationship in a ticket comment instead. Worth adding the real link if you want it.
- This branch is now 1 commit behind `main` after #131 landed. GitHub reports it `MERGEABLE` and the new commit only touches `ManageCookies` and `checks.yml`, neither of which this branch goes near — so I've left it rather than force-pushing a rebase you didn't ask for. Happy to rebase if you'd prefer it current.

Nothing else outstanding: both review threads are resolved and `build` and `Jira` are green on `dcde66a`.

**Review by RoyEJohnson — CHANGES_REQUESTED** [review]
Consider Copilot's latest suggestion

File: src/components/ButtonBar.tsx [hunk 4]
```diff
@@ -1,36 +1,37 @@
...
-  ${tabListBaseCss}
-  ${buttonBarWrapperCss}
-
-  > * {
-    all: unset;
-    ${tabBaseCss}
-    ${buttonBarItemCss}
-  }
-`;
+} & React.HTMLAttributes<HTMLDivElement>;
```
[the remark this answers is not in the document]
Comment (OpenStaxClaude) [reply src/components/ButtonBar.tsx]: Agreed — done in 9891f7f, along with the matching `Tooltip` point from the suppressed comment.

`ButtonBarProps` is now:

```ts
type ButtonBarProps = {
  size?: "large" | "medium" | "small";
  children?: React.ReactNode;
  style?: CSSPropertiesWithVariables;
} & Omit<React.HTMLAttributes<HTMLDivElement>, 'style'>;
```

and `ClassNameAndStyle.style` in `Tooltip.tsx` is widened the same way, which covers `Tooltip`, `StyledTooltip` and `StyledTrigger`.

**I also applied it to `Tabs`,** which wasn't flagged but has the identical problem through `RAC.TabsProps` — its `style` is react-aria's `CSSProperties | ((values) => CSSProperties | undefined)`, so `--tabs-border-color` needed a cast too. Fixing two of the three would have left the API inconsistent, given the PR documents all three variable sets as the theming mechanism.

One thing worth calling out about the `Tabs` change: it uses `Omit<RAC.TabsProps, 'style'>` rather than a plain intersection. Intersecting a narrower `style` over `RAC.TabsProps` is exactly what collapsed the `T | ((values) => T)` union and broke render callbacks earlier in this PR, so the original has to be removed before the wider one goes on. There's a comment in the file saying so, and the existing callback tests guard it.

For `Toast` and `ToastContainer` there was nothing to do — neither exposes a `style` prop at all.

Each component gets a test that overrides a documented variable with no cast. Against the previous types they fail to compile with precisely the problem you described:

```
error TS2353: Object literal may only specify known properties, and
'"--button-bar-border-color"' does not exist in type 'Properties<string | number, string & {}>'.
```

This is a widening, so existing `style` usage is unaffected. Noted in `CHANGELOG.md` under Non-Breaking Changes.

**Review by jivey — APPROVED** [review]

**Review by RoyEJohnson — APPROVED** [review]
